iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp

```diff
@@ -307,6 +307,318 @@ class AdjustDepthwiseFilterShape : public OpRewritePattern<mhlo::ConvOp> {
   }
 };
 
+// Rewrites rank-3 mhlo.dot general in a form where lhs contraction is 2 and rhs
+// contraction dim is 1 and insert required transposes.
+class TransposeRank3GenericDotGeneral
+    : public OpRewritePattern<mhlo::DotGeneralOp> {
+ public:
+  using OpRewritePattern<mhlo::DotGeneralOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::DotGeneralOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhsShapeType = op.lhs().getType().cast<ShapedType>();
+    auto rhsShapeType = op.rhs().getType().cast<ShapedType>();
+    auto resultType = op.getResult().getType().cast<ShapedType>();
+
+    if (!lhsShapeType || !rhsShapeType || !resultType) return failure();
+    if (!lhsShapeType.hasStaticShape() || !rhsShapeType.hasStaticShape())
+      return failure();
+    if (resultType.getRank() != 3) return failure();
+
+    if (op.dot_dimension_numbers().lhs_contracting_dimensions().size() != 1 ||
+        op.dot_dimension_numbers().rhs_batching_dimensions().size() != 1)
+      return failure();
```

Comment:
This check looks weird to me...do we need to check lhs batching dims and rhs contracting dims?